### PR TITLE
fix(sync-actions): fix non-matching variant image update action generation

### PR DESCRIPTION
**Pull request by tdeekens**
#### Summary

This is a bug repo of #305.

### CODE REVIEW

File: packages/sync-actions/src/product-actions.js [hunk 1]
```diff
@@ -177,15 +178,19 @@ export function actionsMapImages (diff, oldObj, newObj) {
   let actions = []
   const { variants } = diff
 
-  if (variants)
+  if (variants) {
+    const variantHashMap = findMatchingPairs(variants, oldObj.variants, newObj.variants, 'id')
```
Comment (tdeekens) [inline packages/sync-actions/src/product-actions.js]: Call it `machingVariantPairs` to make it more descriptive?

File: packages/sync-actions/src/product-actions.js [hunk 2]
```diff
@@ -453,31 +459,38 @@ function _buildVariantImagesAction (diffedImages, oldVariant, newVariant) {
...
             action: 'removeImage',
             variantId: oldVariant.id,
-            imageUrl: oldVariant.images[key].url,
+            imageUrl: oldVariant.images[imagesHashMap[key][0]].url,
           })
           actions.push({
             action: 'addExternalImage',
             variantId: oldVariant.id,
-            image: newVariant.images[key],
+            image: newVariant.images[imagesHashMap[key][1]],
```
Comment (tdeekens) [inline packages/sync-actions/src/product-actions.js]: Can we give this variable a name? `const newImageKey = imagesHashMap[key][1]`?

File: packages/sync-actions/src/product-actions.js [hunk 3]
```diff
@@ -453,31 +459,38 @@ function _buildVariantImagesAction (diffedImages, oldVariant, newVariant) {
...
-          action: 'removeImage',
-          variantId: oldVariant.id,
-          imageUrl: oldVariant.images[index].url,
-        })
-    }
+    } else if (REGEX_UNDERSCORE_NUMBER.test(key))
+      if (Array.isArray(image) && image.length === 3)
+        if (Number(image[2]) === 3)
+          // image position changed
+          actions.push({
```
Comment (tdeekens) [inline packages/sync-actions/src/product-actions.js]: Nice, this wasn't supported back in the days which is why the MC does create their own update actions in the image view.

File: packages/sync-actions/src/utils/find-matching-pairs.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
+import forEach from 'lodash.foreach'
+
+const REGEX_NUMBER = new RegExp(/^\d+$/)
+const REGEX_UNDERSCORE_NUMBER = new RegExp(/^_\d+$/)
+
+function preProcessCollection (variants, identifier) {
+  const refByIndex = {}
+  const refByIdentifier = {}
+  forEach(variants, (variant, index) => {
```
Comment (tdeekens) [inline packages/sync-actions/src/utils/find-matching-pairs.js]: Isn't this more of a `reduce`?

File: packages/sync-actions/test/utils/find-matching-pairs.spec.js [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+        id: 5,
+        sku: 'testing-animation',
+        prices: [],
+        images: [],
+        assets: [],
+      },
+    ]
+  })
+
+  it('should find matching pairs', () => {
```
Comment (tdeekens) [inline packages/sync-actions/test/utils/find-matching-pairs.spec.js]: Lines to first `it` `177`. Not saying it's bad but setup here is really a lot.

File: packages/sync-actions/test/utils/find-matching-pairs.spec.js [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+    const actualResult = findMatchingPairs(
+      diff.variants,
+      oldVariants,
+      newVariants,
+      'id',
+    )
+    const expectedResult = {
+      0: ['3', '0'],
```
Comment (tdeekens) [inline packages/sync-actions/test/utils/find-matching-pairs.spec.js]: Nice!

File: packages/sync-actions/src/utils/find-matching-pairs.js [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    refByIndex[String(index)] = variant[identifier]
+    refByIdentifier[variant[identifier]] = String(index)
+  })
+  return { refByIndex, refByIdentifier }
+}
+
+// creates a hash of a location of an item in collection1 and collection2
+export default function findMatchingPairs (diff, coll1, coll2, identifier) {
```
Comment (tdeekens) [inline packages/sync-actions/src/utils/find-matching-pairs.js]: `coll1` and `coll2` seem like a new name to `before` and `now`?

Comment (tdeekens) [reply packages/sync-actions/src/utils/find-matching-pairs.js]: Should the `identifier` default to `'id'`?

File: packages/sync-actions/src/product-actions.js [hunk 8]
```diff
@@ -204,15 +210,21 @@ export function actionsMapPrices (diff, oldObj, newObj) {
...
-
-      addPriceActions = addPriceActions.concat(a)
-      changePriceActions = changePriceActions.concat(c)
-      removePriceActions = removePriceActions.concat(r)
+      const {
+        oldObj: oldVariant,
+        newObj: newVariant,
+      } = _extractHashMap(variantHashMap, key, oldObj.variants, newObj.variants)
+      if (REGEX_UNDERSCORE_NUMBER.test(key) || REGEX_NUMBER.test(key)) {
+        const [ a, c, r ] = _buildVariantPricesAction(
```
Comment (tdeekens) [inline packages/sync-actions/src/product-actions.js]: ...why don't we call them `additionalAddedPriceActions` etc?

Comment (hisabimbola) [reply packages/sync-actions/src/product-actions.js]: that's a good questions, met it like that, but it can be improved

**Comment by hisabimbola** [comment]
Since I took over this, doesn't make sense to approve my own code and @tdeekens can't approve the code because he opened the PR 🙄 . @tdeekens you can add a comment if it's ready and I will approve and merge.

**Comment by hisabimbola** [comment]
Though we did a major refactor I think this should still be released as a `patch` update right. Let me know if anyone things otherwise

File: packages/sync-actions/src/product-actions.js [hunk 9]
```diff
@@ -434,10 +451,26 @@ function _buildSetAttributeAction (
...
-function _buildVariantImagesAction (diffedImages, oldVariant, newVariant) {
+function _buildVariantImagesAction (
+  diffedImages,
+  oldVariant = {},
+  newVariant = {},
+) {
   const actions = []
-
+  // generate a hashMap to be able to reference the right image from both ends
+  const imagesHashMap = findMatchingPairs(
```
Comment (tdeekens) [inline packages/sync-actions/src/product-actions.js]: Can we just call this `matchingImagePairs` to be descriptive around what's inside not what it is?

File: packages/sync-actions/src/product-actions.js [hunk 10]
```diff
@@ -454,42 +487,85 @@ function _buildVariantImagesAction (diffedImages, oldVariant, newVariant) {
...
+          })
+      }
   })
 
   return actions
 }
 
-function _buildVariantPricesAction (diffedPrices, oldVariant, newVariant) {
+// safely extract oldObj and newObj
+function _extractHashMap (hashMap, key, before, now) {
```
Comment (tdeekens) [inline packages/sync-actions/src/product-actions.js]: _extractMatchingNewAndOld?

File: packages/sync-actions/src/product-actions.js [hunk 11]
```diff
@@ -454,42 +487,85 @@ function _buildVariantImagesAction (diffedImages, oldVariant, newVariant) {
...
+  diffedPrices,
+  oldVariant = {},
+  newVariant = {},
+) {
   const addPriceActions = []
   const changePriceActions = []
   const removePriceActions = []
 
+  // generate a hashMap to be able to reference the right image from both ends
+  const pricesHashMap = findMatchingPairs(
```
Comment (tdeekens) [inline packages/sync-actions/src/product-actions.js]: matchingPricePairs?

**Comment by hisabimbola** [comment]
Hey @tdeekens thanks for your reviews, I have updated the code.

about the `new`, `old`, `before` or `now` they all mean the same thing, but I doubt we use `new` at all because it's a *reserved* keyword in JS.

**Comment by hisabimbola** [comment]
If we merge this, it will be publish straight away, we don't have custom tag implemented.

**Comment by hisabimbola** [comment]
and yeah, I got the base test case from #305, so I will safely assume that this PR should fix the issue.

**Comment by tdeekens** [comment]
Let's :shipit: and we'll test it right away. @hisabimbola you can do the honors.

**Pull request merged**